Adding ZMQ/Pub support for txpool_add and chain_main events

vtnerd opened the pull request:
Supports txpool and mainchain events. Will be updated after PR 6357 is merged. Markdown file has description on how clients should use it, and code has decent comments on implementation quirks.

moneromooo-monero commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -1015,39 +1037,48 @@ namespace cryptonote
+      {MERROR_VER("Transaction verification failed: " << results[i].tx.hash);}
       else if(tvc[i].m_verifivation_impossible)
-      {MERROR_VER("Transaction verification impossible: " << results[i].hash);}
+      {MERROR_VER("Transaction verification impossible: " << results[i].tx.hash);}
```
While these changes seem OK in practice, tx.hash is not really meant to be public. It's a cache that's only valid is hash_valid is set. Encouraging this use instead of get_transaction_hash will likely bite us in the rear later. So this adds a couple dozen mostly pointless lines and makes the code more brittle IMHO, even though correct.

vtnerd replied:
Then why is the field public? `get_transaction_hash` should be a public member function that checks the private internal variable and then forwards to the global function. Otherwise, this is clearly indicating accessibility to all. 

This would seem to only "bite us" if one of these functions modifies the hash after its already been calculated once. Seems like a weird update, but yeah sure I'll just duplicate it in memory like before.

vtnerd commented on ZMQ.md:
```diff
@@ -0,0 +1,61 @@
+ZMQ Pub/Sub will drop messages if the network is congested, so the above rules
+for send order are used for detecting lost messages. A missing gap in `height`
+or `prev_id` for `chain_*` events indicates a lost pub message. Missing
+`txpool_add` messages can only be detected at the next `chain_` message.
```
This is the behavior of ZMQ pub/sub. There isn't a good method to fix or change it. If a downstream clients fails to read any pub messages, the server is then forced to block/wait or queue an infinite amount of messages. The number of messages should be fairly low for our purposes, so its not expected to happen frequently if at all.

vtnerd commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -262,6 +264,13 @@ namespace cryptonote
+  {
+    CRITICAL_REGION_LOCAL(m_incoming_tx_lock);
+    m_zmq_pub = std::move(zmq_pub);
+  }
```
I can create a new library for `zmq_pub.cpp` with the core and blockchain dependent on that instead. A callback isn't going to work as desired, because the code still needs to be invoked. Those other libs need to know where to jump either way.

vtnerd commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -968,44 +989,45 @@ namespace cryptonote
             MERROR_VER("Exception in handle_incoming_tx_post: " << e.what());
             tvc[i].m_verifivation_failed = true;
-            results[i].res = false;
+            results[i].valid = false;
```
I don't think its pointless, `res` is not descriptive for what the field represents. The change is pretty simple to revert though.

vtnerd replied:
Reverted.

vtnerd commented:
Rebased, and changes from resolved comments.

selsta commented:
@vtnerd Please rebase again.

vtnerd commented:
Rebased.

vtnerd commented on src/rpc/zmq_pub.cpp:
```diff
@@ -0,0 +1,462 @@
+
+  void json_full_chain(rapidjson::StringBuffer& buf, const std::uint64_t height, const cryptonote::block& bl)
+  {
+    json_pub(buf, std::cref(bl));
```
The JSON schema only allows one block at a time. When the daemon is synchronizing, this sends lots of individual ZMQ messages. Actually changing this will require a bit more work in the "core" code, but the schema doesn't allow any flexibility right now.

I'm thinking about making the schema:
```json
{"first_height": ..., "last_id": ..., "blocks: [ ... ]}
```
I initially decided against this for micro-optimization purposes on the receiver side - they need the equivalent of a `std::vector` which is less efficient for the more common single block case. However, this can be mitigated in several different ways (`boost::container::small_vector`, etc).

The other advantage is that the full block send doesn't contain the hash directly right now (it must be computed by duplicating what the daemon does), and getting the block height for full block requires poking into the `miner_tx`. So these fields are redundant for the full block, but still kind of helpful due to the complexity of the schema.

Any NACKs?

vtnerd replied:
I switched to arrays for `chain_main` events (future extensibility), but did not change the full block output. Minimal block events have differed slightly with array change.

vtnerd commented on src/rpc/zmq_pub.cpp:
```diff
@@ -0,0 +1,462 @@
+    namespace adapt = boost::adaptors;
+    const auto to_full_tx = [](const cryptonote::txpool_event& event)
+    {
+      return event.tx;
```
It just occurred to me that the full txes don't have the hash included either. They must be computed remotely by duplicating the daemon code. Worth sending the hash here too?

vtnerd commented:
This has been updated to use the new `byte_slice` send feature with ZMQ that was recently merged. Should've done this during the rebase, but the existing code worked without it.

xiphon commented on src/rpc/CMakeLists.txt:
```diff
@@ -83,6 +87,8 @@ set(daemon_rpc_server_private_headers
 monero_private_headers(rpc
   ${rpc_private_headers})
 
+set(rpc_pub_private_headers)
```
If it is empty, why we need it here?

vtnerd replied:
This is the style in this file. The empty set is declared like this and passed to the `monero_add_library` function. I can revert to `""` is desired, I only did it this way to follow the existing style.

xiphon commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -1044,10 +1044,13 @@ bool Blockchain::switch_to_alternative_blockchain(std::list<block_extended_info>
-    for (const auto &bei: alt_chain)
-      block_notify->notify("%s", epee::string_tools::pod_to_hex(get_block_hash(bei.bl)).c_str(), NULL);
+  std::size_t notify_height = split_height;
+  for (const auto& bei: alt_chain)
+  {
+    for (const auto& notifier: m_block_notifiers)
```
Consider to swap the loops

vtnerd replied:
Any more information? I wrote it this way because the same block should remain in the CPU data cache for each call. However, this should thrash the CPU instruction cache. Are you saying you think the opposite is preferred? Some other reason?

xiphon replied:
In case of empty `m_block_notifiers`, we could skip iteration over `alt_chain`.

vtnerd replied:
Done.

xiphon commented on src/cryptonote_core/CMakeLists.txt:
```diff
@@ -59,6 +59,7 @@ target_link_libraries(cryptonote_core
     ringct
     device
     hardforks
+    rpc_pub
```
Don't like this. IMO we should avoid introducing this dependency.

moneromooo-monero replied:
Agreed. If it's a pain to avoid I could live with it though.

vtnerd replied:
Using `boost::function` (not `std::function`, see comments). The penalty is an indirect function call instead of a direct function call when `zmq_pub` is enabled. Its also less obvious now  that the call is doing a `shared_ptr` lock, but the code for blockchain notification has the same issue.

vtnerd commented:
* Rebased
* `chain_main` now sends block information in arrays
* Changed the serialization routine (necessary for rebase) to allocate memory more efficiently
* Cryptonote core does not depend on ZMQ-PUB
* Fixed bug in `byte_slice` (another PR coming shortly).

moneromooo-monero commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -1016,21 +1038,23 @@ namespace cryptonote
         ok = false;
         continue;
       }
+      results[i].valid = false;
```
Why ? Looks like you set valid to false for txes we alreay have in the txpool or the chain, and therefore are valid.

vtnerd replied:
This is probably the result of several edits, and is now confusing. I will remove this line and 1074, then add an `else { results[i].valid = false }` section below the `if` block that starts on line 1071. The intent was to make sure that txes _not_ added to the pool are marked as `!invalid` so that they are not reported over ZMQ (the ZMQ notify code skips those txes).

vtnerd commented:
* Rebased
* `valid` -> `res` (revert name change)
* Updating `if`/`else` block to be more clear.

vtnerd commented:
Ok, pushed the changes forreal this time.

moneromooo-monero commented:
I'm going to assume these are the only changes, and not actually re-read due to the huge rebase diffs.
